icu4c/source/test/intltest/unitstest.cpp

```diff
@@ -242,10 +250,274 @@ void UnitsTest::testConversions() {
-    if (errorCode.errIfFailureAndReset("error parsing %s: %s\n", filename, u_errorName(errorCode))) {
+    if (errorCode.errIfFailureAndReset("error parsing %s: %s\n", path.data(), u_errorName(errorCode))) {
+        return;
+    }
+}
+
+/**
+ * This class represents the output fields from unitPreferencesTest.txt. Please
+ * see the documentation at the top of that file for details.
+ *
+ * For "mixed units" output, there are more (repeated) output fields. The last
+ * output unit has the expected output specified as both a rational fraction and
+ * a decimal fraction. This class ignores rational fractions, and expects to
+ * find a decimal fraction for each output unit.
+ */
+class ExpectedOutput {
+  private:
+    // Counts number of units in the output. When this is more than one, we have
+    // "mixed units" in the expected output.
+    int _compoundCount = 0;
+
+    // Counts how many fields were skipped: we expect to skip only one per
+    // output unit type (the rational fraction).
+    int _skippedFields = 0;
+
+    // The expected output units: more than one for "mixed units".
+    MeasureUnit _measureUnits[3];
+
+    // The amounts of each of the output units.
+    double _amounts[3];
+
+  public:
+    /**
+     * Parse an expected output field from the test data file.
+     * @param output may be a string representation of an integer, a rational
+     * fraction, a decimal fraction, or it may be a unit identifier. Whitespace
+     * should already be trimmed. This function ignores rational fractions,
+     * saving only decimal fractions and their unit identifiers.
+     * @return true if the field was successfully parsed, false if parsing
+     * failed.
+     */
+    UBool parseOutputField(StringPiece output) {
+        DecimalQuantity dqOutputD;
+        UErrorCode errorCode = U_ZERO_ERROR;
+
+        dqOutputD.setToDecNumber(output, errorCode);
+        if (U_SUCCESS(errorCode)) {
+            _amounts[_compoundCount] = dqOutputD.toDouble();
+            return true;
+        } else if (errorCode == U_DECIMAL_NUMBER_SYNTAX_ERROR) {
+            errorCode = U_ZERO_ERROR;
+        } else {
+            return false;
+        }
+
+        _measureUnits[_compoundCount] = MeasureUnit::forIdentifier(output, errorCode);
+        if (U_SUCCESS(errorCode)) {
+            _compoundCount++;
+            _skippedFields = 0;
+            return true;
+        }
+        _skippedFields++;
+        if (_skippedFields < 2) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Produces an output string for debug purposes.
+     * @param dbgOutput the buffer to which to write the output string.
+     * @param dbgOutputSize the size of the buffer. 
+     * @returns a null-terminated string containing each amount and unit.
+     */
+    void toDebugStringN(char *dbgOutput, int dbgOutputSize) {
```

Comment:
Thanks! I'll try to remember that (I made a note). For this code, I went with std:string and std::to_string, because these identifiers are all simple ASCII.